source/test/TestApp/PlayFabEventTest.cpp

```diff
@@ -91,12 +91,23 @@ namespace PlayFabUnit
     int PlayFabEventTest::eventPassCount;
     int PlayFabEventTest::eventFailCount;
     std::string PlayFabEventTest::eventFailLog;
+    unsigned int PlayFabEventTest::numEventsHeard;
+    std::vector<std::thread> PlayFabEventTest::testThreadPool;
 
     void PlayFabEventTest::NonStaticEmitEventCallback(std::shared_ptr<const PlayFab::IPlayFabEvent> /*event*/, std::shared_ptr<const PlayFab::IPlayFabEmitEventResponse> /*response*/)
     {
         (*eventTestContext)->Pass("Private member called back!");
     }
 
+    void PlayFabEventTest::MyThreadingCallback(std::shared_ptr<const PlayFab::IPlayFabEvent> /*event*/, std::shared_ptr<const PlayFab::IPlayFabEmitEventResponse> /*response*/)
+    {
+        numEventsHeard++;
+        if(numEventsHeard == numTotalThreadedEvents)
```

Comment:
>numTotalThreadedEvents [](start = 29, length = 22)

only one test ever (intentionally) expects this exact number of events. The other two tests don't have any relation to this constant and only work because they are firing greater than or equal to this number of events. I would consider this another example where reliance on shared static state has led to less clear and semantically incorrect code. I think there's a better way to do what you are trying to do using condition variables. Let me see if I can jot down my idea in some quick pseudo because I'm having difficulty putting it in non-code words

Reply:
here's some pseudo-ish code

```
void GenericMultiThreadedTest(pNumThreads, pNumEventsPerThread)
{
    bool testDone = false;
    uint32_t totalEventsToWaitFor = pNumThreads * pNumEventsPerThread;

    auto EmitEventCallback = [<captured state>](<callbackParams>)
    {
        // Take special care to make sure multiple threads can't write this at the same time
        totalEventsToWaitFor -= 1;
        if (totalEventsToWaitFor == 0)
        {
            PassTest()!
        }
    }

    for (unsigned int thread = 0; thread < pNumThreads; ++thread)
    {
        testThreadPool.emplace_back(
            [pNumEventsPerThread, &EmitEventCallback]()
            {
                std::shared_ptr<PlayFabEventAPI*> api = SetupEventTest();
                for (unsigned int i = 0; i < pNumEventsPerThread; ++i)
                {
                    (*api)->EmitEvent(MakeEvent(0, PlayFabEventType::Default), EmitEventCallback);
                }
            });
    }

    {
        // Wait for all events to fire
        // TODO:
    }

    for (thread : testThreadPool)
    {
        thread.join();
    }
    testThreadPool.clear();
}
```

I didn't bother writing out any of the synchronization bits (figured since I wasn't writing the code and wouldn't bother verifying what I wrote those kinds of details would best be left as an exercise for you) but those are really important and must be done.

anyway I think this pseudo accomplishes removing a bunch of static state and making all of the relevant test code available in one Test function so you don't have to jump around the file to understand what happening

---
In reply to: [345442525](https://github.com/PlayFab/XPlatCoreTemplate/pull/44#discussion_r345442525) [](ancestors = 345442525)